File: frontend/src/worker/shared/FormulaParser.ts
// ROLE: Парсер формул роста. ЕДИНСТВЕННАЯ копия — импортируется только воркером.

const tokenCache = new Map<string, string[]>();

export class FormulaParser {
  private tokens: string[] = [];
  private pos = 0;

  constructor(expression: string, s: number, l: number) {
    if (!tokenCache.has(expression)) {
      const regex = /\d+(?:\.\d+)?|[a-z_][a-z0-9_]*|[\+\-\*\/\^,\(\)]/gi;
      tokenCache.set(expression, expression.match(regex) || []);
    }
    this.tokens = [...tokenCache.get(expression)!];
    this.pos = 0;
    
    for (let i = 0; i < this.tokens.length; i++) {
      const t = this.tokens[i].toLowerCase();
      if (t === "s" || t === "score") {
        this.tokens[i] = String(s);
      } else if (t === "l" || t === "len" || t === "length") {
        this.tokens[i] = String(l);
      } else if (t === "pi") {
        this.tokens[i] = String(Math.PI);
      } else if (t === "e") {
        this.tokens[i] = String(Math.E);
      }
    }
  }

  private peek(): string | null {
    return this.pos < this.tokens.length ? this.tokens[this.pos] : null;
  }

  private consume(expected?: string): string {
    const token = this.peek();
    if (!token) throw new Error("Unexpected end of expression");
    if (expected && token !== expected) throw new Error(`Expected token ${expected}`);
    this.pos++;
    return token;
  }

  public parse(): number {
    try {
      const val = this.parseExpression();
      if (this.peek() !== null) throw new Error("Extra tokens at end");
      return isNaN(val) ? 10.0 : val;
    } catch {
      return 10.0;
    }
  }

  private parseExpression(): number {
    let val = this.parseTerm();
    while (true) {
      const op = this.peek();
      if (op === "+" || op === "-") {
        this.consume();
        const rhs = this.parseTerm();
        val = op === "+" ? val + rhs : val - rhs;
      } else {
        break;
      }
    }
    return val;
  }

  private parseTerm(): number {
    let val = this.parseFactor();
    while (true) {
      const op = this.peek();
      if (op === "*" || op === "/") {
        this.consume();
        const rhs = this.parseFactor();
        val = op === "*" ? val * rhs : val / (rhs === 0 ? 0.001 : rhs);
      } else {
        break;
      }
    }
    return val;
  }

  private parseFactor(): number {
    let val = this.parsePrimary();
    while (this.peek() === "^") {
      this.consume();
      const exponent = this.parsePrimary();
      val = Math.pow(val, exponent);
    }
    return val;
  }

  private parsePrimary(): number {
    const token = this.consume();
    if (token === "(") {
      const val = this.parseExpression();
      this.consume(")");
      return val;
    }
    if (token === "-") return -this.parsePrimary();
    if (token === "+") return this.parsePrimary();

    const num = Number(token);
    if (!isNaN(num)) return num;

    return this.parseFunction(token.toLowerCase());
  }

  private parseFunction(func: string): number {
    if (!["sin", "cos", "tan", "log", "log10", "sqrt", "abs", "exp", "min", "max", "pow"].includes(func)) {
      return 0.0;
    }
    this.consume("(");
    const args: number[] = [];
    args.push(this.parseExpression());
    while (this.peek() === ",") {
      this.consume();
      args.push(this.parseExpression());
    }
    this.consume(")");

    switch (func) {
      case "sin": return Math.sin(args[0]);
      case "cos": return Math.cos(args[0]);
      case "tan": return Math.tan(args[0]);
      case "log": return Math.log(Math.max(0.001, args[0]));
      case "log10": return Math.log10(Math.max(0.001, args[0]));
      case "sqrt": return Math.sqrt(Math.max(0.0, args[0]));
      case "abs": return Math.abs(args[0]);
      case "exp": return Math.exp(args[0]);
      case "min": return Math.min(...args);
      case "max": return Math.max(...args);
      case "pow": return Math.pow(args[0], args[1] || 0);
    }
    return 0.0;
  }
}

export function evaluateFormula(formula: string | number, score: number, length: number): number {
  if (typeof formula === "number") return formula;
  if (!formula || formula === "") return 10.0;
  const parser = new FormulaParser(formula, score, length);
  const result = parser.parse();
  return Math.max(0.1, result);
}

File: backend/app/engine/systems/formula_parser.py
# ROLE: Парсер формул роста. Точная копия логики из FormulaParser.ts.
import re
import math


_TOKEN_CACHE = {}


class FormulaParser:
    def __init__(self, expression: str, s: float, l: float):
        if expression not in _TOKEN_CACHE:
            regex = r"\d+(?:\.\d+)?|[a-z_][a-z0-9_]*|[\+\-\*\/\^,\(\)]"
            _TOKEN_CACHE[expression] = re.findall(regex, expression, re.IGNORECASE)
        
        self.tokens = list(_TOKEN_CACHE[expression])
        self.pos = 0

        for i in range(len(self.tokens)):
            t = self.tokens[i].lower()
            if t in ("s", "score"):
                self.tokens[i] = str(s)
            elif t in ("l", "len", "length"):
                self.tokens[i] = str(l)
            elif t == "pi":
                self.tokens[i] = str(math.pi)
            elif t == "e":
                self.tokens[i] = str(math.e)

    def peek(self) -> str:
        return self.tokens[self.pos] if self.pos < len(self.tokens) else None

    def consume(self, expected: str = None) -> str:
        token = self.peek()
        if token is None:
            raise ValueError("Unexpected end of expression")
        if expected and token != expected:
            raise ValueError(f"Expected token {expected}")
        self.pos += 1
        return token

    def parse(self) -> float:
        try:
            val = self.parse_expression()
            if self.peek() is not None:
                raise ValueError("Extra tokens at end")
            if math.isnan(val):
                return 10.0
            return val
        except Exception:
            return 10.0

    def parse_expression(self) -> float:
        val = self.parse_term()
        while True:
            op = self.peek()
            if op in ("+", "-"):
                self.consume()
                rhs = self.parse_term()
                val = val + rhs if op == "+" else val - rhs
            else:
                break
        return val

    def parse_term(self) -> float:
        val = self.parse_factor()
        while True:
            op = self.peek()
            if op in ("*", "/"):
                self.consume()
                rhs = self.parse_factor()
                if op == "*":
                    val *= rhs
                else:
                    val /= 0.001 if rhs == 0 else rhs
            else:
                break
        return val

    def parse_factor(self) -> float:
        val = self.parse_primary()
        while self.peek() == "^":
            self.consume()
            exponent = self.parse_primary()
            val = math.pow(val, exponent)
        return val

    def parse_primary(self) -> float:
        token = self.consume()
        if token == "(":
            val = self.parse_expression()
            self.consume(")")
            return val
        if token == "-":
            return -self.parse_primary()
        if token == "+":
            return self.parse_primary()

        try:
            num = float(token)
            return num
        except ValueError:
            pass

        return self.parse_function(token.lower())

    def parse_function(self, func: str) -> float:
        if func not in (
            "sin",
            "cos",
            "tan",
            "log",
            "log10",
            "sqrt",
            "abs",
            "exp",
            "min",
            "max",
            "pow",
        ):
            return 0.0

        self.consume("(")
        args = []
        args.append(self.parse_expression())
        while self.peek() == ",":
            self.consume()
            args.append(self.parse_expression())
        self.consume(")")

        if func == "sin":
            return math.sin(args[0])
        elif func == "cos":
            return math.cos(args[0])
        elif func == "tan":
            return math.tan(args[0])
        elif func == "log":
            return math.log(max(0.001, args[0]))
        elif func == "log10":
            return math.log10(max(0.001, args[0]))
        elif func == "sqrt":
            return math.sqrt(max(0.0, args[0]))
        elif func == "abs":
            return abs(args[0])
        elif func == "exp":
            return math.exp(args[0])
        elif func == "min":
            return min(args)
        elif func == "max":
            return max(args)
        elif func == "pow":
            return math.pow(args[0], args[1] if len(args) > 1 else 0)

        return 0.0


def evaluate_formula(formula, score: float, length: float) -> float:
    if isinstance(formula, (int, float)):
        return float(formula)
    if not formula:
        return 10.0
    parser = FormulaParser(str(formula), score, length)
    result = parser.parse()
    return max(0.1, result)
